sw/host/ot_certs/src/offsets.rs

```diff
@@ -0,0 +1,195 @@
+}
+
+struct GeneratorVariable {
+    hidden: bool,
+    source_type: VariableType,
+    values: Vec<Vec<u8>>,
+    offsets: Vec<CertificateOffset>,
+}
+
+impl GeneratorVariable {
+    fn new(source_type: VariableType) -> GeneratorVariable {
+        GeneratorVariable {
+            hidden: false,
+            source_type,
+            values: Vec::new(),
+            offsets: Vec::new(),
+        }
+    }
+}
+
+// The generators holds the information about the offset to the variables
+// in the certificate and the randomly generated values that we used during generation.
+pub struct OffsetGenerator {
+    variables: HashMap<String, GeneratorVariable>,
+}
+
+// Super inefficient function to search for needle in a haystack.
+fn find_subsequence(haystack: &[u8], needle: &[u8]) -> Option<usize> {
+    haystack
+        .windows(needle.len())
+        .position(|window| window == needle)
+}
+
+impl OffsetGenerator {
+    pub fn new(variables: &HashMap<String, VariableType>) -> OffsetGenerator {
+        let variables = variables
+            .iter()
+            .map(|(name, source_type)| (name.clone(), GeneratorVariable::new(*source_type)))
+            .collect();
+        OffsetGenerator { variables }
+    }
+
+    pub fn add_variable<T>(&mut self, bytes: Vec<u8>, target_type: VariableType, val: &Value<T>) {
+        let Value::Variable { var: name, convert } = val else {
+            panic!("add_variable can only be called on variables, not literals");
+        };
+        // Register value.
+        if let Some(entry) = self.variables.get_mut(name) {
+            entry.values.push(bytes);
+            // Register offset.
+            entry.offsets.push(CertificateOffset {
+                offset: 0,
+                target_type,
+                conversion: *convert,
+            });
+        } else {
+            // This should never occur because the variable name should have been checked before
+            // in get_value.
+            panic!("variable '{}' does not exist", name);
+        }
+    }
+
+    pub fn add_hidden_variable(&mut self, name: String, bytes: Vec<u8>) {
+        let target_type = VariableType::ByteArray { size: bytes.len() };
+        let off = CertificateOffset {
+            offset: 0,
+            target_type,
+            conversion: None,
+        };
+        let var = GeneratorVariable {
+            hidden: true,
+            source_type: target_type,
+            values: vec![bytes],
+            offsets: vec![off],
+        };
+        self.variables.insert(name, var);
+    }
+
+    fn compute_offsets(
+        &self,
+        cert: &mut [u8],
+        clear_variables: bool,
+    ) -> Result<Vec<CertificateVariable>> {
+        let mut res = Vec::new();
+        for (name, var) in self.variables.iter() {
+            let mut offsets = Vec::new();
+            for (i, offset) in var.offsets.iter().enumerate() {
+                let needle = &var.values[i];
+                match find_subsequence(cert, needle) {
```

Comment:
IIUC for variables we generate random sequence and then identify offsets using these random sequence?

In that case, I think it's worth having a sanity check to ensure that there're no multiple occurrence of the subsequence just in case.

Reply:
Yes. Yeah that makes sense.